### PR TITLE
[WIP][Syntax] Performance improvements

### PR DESCRIPTION
**Pull request by rintaro**
**⚠️  DO NOT MERGE**

* Don't construct complete `Syntax` nodes in parsing.
  Currently, there're many `RC<RawSyntax>` → `Syntax` → `RC<RawSyntax>` conversions. Since `Syntax` nodes have `RC<SyntaxData>` that needs heap allocation, frequent conversions cost a lot. Instead, directly use `RC<RawSyntax>`. #14297 
* Represent missing optional nodes as `nullptr`.
  For instance, every `IdentifierExpr` have optional `DeclNameArguments` (e.g. `(x:y:)`) node. Allocating it for every identifier expressions is definitely a waste of time and resource. #14300 
* Cache and reuse `RawSyntax` nodes (particularly, tokens) when possible.
  For instance, in `foo(), bar()`, both `(` have exact the same layout of `RawSyntax` (kind, text and leading/trailing trivia). We should be able to reuse one `RawSyntax` tokens to save allocation costs.
  Currently, the cache storage is in `ASTContext`. Probably we should create this kind of infrastructure for `libSyntax`. #14330
* Use `TrailingObjects` for `SyntaxData`.
  This should optimize memory allocations. #14301
* Use bump-pointer allocation for `RawSyntax`.
  Heap allocation and thread safe reference counting are slow. #14330

### CODE REVIEW

File: test/Syntax/Inputs/serialize_multiple_decls.json [hunk 1]
```diff
@@ -17,18 +17,10 @@
                       "kind": "StructDecl",
                       "layout": [
                         {
-                          "kind": "AttributeList",
```
Comment (harlanhaskins) [inline test/Syntax/Inputs/serialize_multiple_decls.json]: Big fan of the JSON space savings this wins us! Hopefully the Codable changes to `RawSyntax.swift` aren't too crazy, but I absolutely am a fan of avoiding serializing the contents of missing nodes altogether.

**Comment by nkcsgexi** [comment]
Nice, this doesn't only provide us performance gain on the compiler side but also the Json deserialization on the Swift side.

**Comment by harlanhaskins** [comment]
I’d also like to turn off JSON pretty printing. White space is a majority of the JSON size.

File: test/Syntax/Inputs/serialize_multiple_decls.json [hunk 2]
```diff
@@ -16,20 +16,8 @@
...
-                          "presence": "Missing"
-                        },
-                        {
-                          "kind": "DeclModifier",
-                          "layout": [
-
-                          ],
-                          "presence": "Missing"
-                        },
+                        null,
```
Comment (rintaro) [inline test/Syntax/Inputs/serialize_multiple_decls.json]: @harlanhaskins I'm going to use `null` here. Fixing `SwiftSyntax.RawSyntax`...

```swift
enum RawSyntax: Codable {
  /// A tree node with a kind, an array of children, and a source presence.
  case node(SyntaxKind, [RawSyntax?], SourcePresence)
                         ~~~~~~~~~^
```

**Pull request closed without merging**